nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/AbstractEmailProcessor.java

```diff
@@ -0,0 +1,391 @@
+     */
+    protected abstract T buildMessageReceiver(ProcessContext context);
+
+    /**
+     * Return the target receivere's mail protocol (e.g., imap, pop etc.)
+     */
+    protected abstract String getProtocol(ProcessContext processContext);
+
+    /**
+     * Builds the url used to connect to the email server.
+     */
+    String buildUrl(ProcessContext processContext) {
+        String host = processContext.getProperty(HOST).evaluateAttributeExpressions().getValue();
+        String port = processContext.getProperty(PORT).evaluateAttributeExpressions().getValue();
+        String user = processContext.getProperty(USER).evaluateAttributeExpressions().getValue();
+        String password = processContext.getProperty(PASSWORD).evaluateAttributeExpressions().getValue();
+        String folder = processContext.getProperty(FOLDER).evaluateAttributeExpressions().getValue();
+
+        StringBuilder urlBuilder = new StringBuilder();
+        try {
+            urlBuilder.append(URLEncoder.encode(user, "UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            throw new ProcessException(e);
+        }
+        urlBuilder.append(":");
+        try {
+            urlBuilder.append(URLEncoder.encode(password, "UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            throw new ProcessException(e);
+        }
+        urlBuilder.append("@");
+        urlBuilder.append(host);
+        urlBuilder.append(":");
+        urlBuilder.append(port);
+        urlBuilder.append("/");
+        urlBuilder.append(folder);
+
+        String protocol = this.getProtocol(processContext);
+        String finalUrl = protocol + "://" + urlBuilder.toString();
+
+        // build display-safe URL
+        int passwordStartIndex = urlBuilder.indexOf(":") + 1;
+        int passwordEndIndex = urlBuilder.indexOf("@");
+        urlBuilder.replace(passwordStartIndex, passwordEndIndex, "[password]");
+        this.displayUrl = protocol + "://" + urlBuilder.toString();
+        if (this.logger.isInfoEnabled()) {
+            this.logger.info("Connecting to Email server at the following URL: " + this.displayUrl);
+        }
+
+        return finalUrl;
+    }
+
+    /**
+     * Builds and initializes the target message receiver if necessary (if it's
+     * null). Upon execution of this operation the receiver is fully functional
+     * and is ready to receive messages.
+     */
+    private synchronized void initializefNecessary(ProcessContext context, ProcessSession processSession) {
+        if (this.messageReceiver == null) {
+            this.processSession = processSession;
+            this.messageReceiver = this.buildMessageReceiver(context);
+
+            boolean shouldDelete = context.getProperty(SHOULD_DELETE_MESSAGES).asBoolean();
+            int fetchSize = context.getProperty(FETCH_SIZE).evaluateAttributeExpressions().asInteger();
+
+            this.messageReceiver.setShouldDeleteMessages(shouldDelete);
```

Comment:
I haven't tested fully, but am I correct to assume that Spring agent will delete the message upon retrieval, independent of the flow file being generated or not?

If that is the case, wouldn't this lead to message loss?

Wouldn't be safer to use the fetchMessages method followed by a deleteMessages one the message is sent to success?